### PR TITLE
FTX: Margin lending/borrow rate history

### PR DESCRIPTION
**Pull request by gloriousCode**
# PR Description
Adds the ability to get FTX margin lending/borrow rate history, check your borrow costs/lending payments and get a nice summary. Also adds support in GRPC and the GCTCLI. The CLI support is partial, since offline calculation requires rates, there's no friendly way of allowing someone to type in rates manually via CMD. The tests under rpcserver_tests.go highlight that it works.

This will be a nice addition for traders looking to see how much it is costing them to be in a leveraged spot margin position

```
go run . getmarginrateshistory --exchange="ftx" --asset="spot" --currency="usd" --bc --br --lp -p
{
 "total_rates": 745,
 "sum_borrow_costs": "0",
 "sum_borrow_size": "0",
 "sum_lending_payments": "0",
 "sum_lending_size": "0",
 "latest_rate": {
  "time": "2022-06-21 05:00:00 +0000",
  "rate": "0.00000228",
  "yearly_rate": "0.0199728",
  "borrow_rate": "0.000003078"
 },
 "predicted_rate": {
  "time": "2022-06-21 06:00:00 +0000",
  "rate": "0.00000228",
  "yearly_rate": "0.0199728",
  "borrow_rate": "0.000003078"
 },
 "taker_fee_rate": "0.0007"
}
```
image

## Type of change

- [x] New feature (non-breaking change which adds functionality)

## How has this been tested
- [x] go test ./... -race
- [x] golangci-lint run
- [x] buf format -w
- [x] Coverage of all new functionality

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation and regenerated documentation via the documentation tool
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally and on Github Actions/AppVeyor with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: cmd/gctcli/commands.go [hunk 1]
```diff
@@ -5060,3 +5060,202 @@ func shutdown(c *cli.Context) error {
...
+			Value:       time.Now().Format(common.SimpleTimeFormat),
+			Destination: &endTime,
+		},
+		&cli.BoolFlag{
+			Name:    "getpredictedrate",
+			Aliases: []string{"p"},
+			Usage:   "include the predicted upcoming rate in the response",
+		},
+		&cli.BoolFlag{
+			Name:    "getlendingpayments",
```
Comment (shazbert) [inline cmd/gctcli/commands.go:5105]:
```
 go run . getmarginrateshistory -e='ftx' -a='spot' -c='btc' -br -bc --lp
```
 "sum_lending_payments": "0.10386594827889",
 "sum_lending_size": "2.46127894827889", 

The lending size of the account is 0.00763200 the above values seem to be exaggerated somewhat.

Doing verbose: 
image

The size randomly jumps up. 🤷

**Review by shazbert — CHANGES_REQUESTED** [review]
Looks really good! Just some minor niterinos and testing things.

File: engine/rpcserver.go [hunk 2]
```diff
@@ -4575,3 +4575,179 @@ func (s *RPCServer) Shutdown(_ context.Context, _ *gctrpc.ShutdownRequest) (*gct
...
+		if err != nil {
+			return nil, err
+		}
+	}
+	err = common.StartEndTimeCheck(start, end)
+	if err != nil {
+		return nil, err
+	}
+
+	request := &order.MarginRateHistoryRequest{
```
[the remark this answers is not in the document]
Comment (gloriousCode) [reply engine/rpcserver.go:4627]: Yeah I felt somewhat unsure of it too. There will be a point in future where one could get a sum of the borrow costs associated with a position, but those borrow costs aren't themselves an order. So yeah, I'll move it to exchanges\margin 👍

File: cmd/gctcli/commands.go (hunk 1 above)
Comment (gloriousCode) [reply cmd/gctcli/commands.go:5105]: Definitely no copy pasting was involved in any errors 👀  https://github.com/thrasher-corp/gocryptotrader/pull/981/files#diff-e7811fdafb5e931926e53a40cf51d3205cfc00947aa56c5e0bac7635e88ad6f7R1828

**Review by shazbert — APPROVED** [review]
tACK, thanks for fixing those nits.

**Review by shazbert — APPROVED** [review]
Changes tested ACK.

File: engine/rpcserver.go [hunk 3]
```diff
@@ -4749,3 +4750,180 @@ func (s *RPCServer) GetTechnicalAnalysis(ctx context.Context, r *gctrpc.GetTechn
...
+			HourlyRate: lendingResp.PredictedRate.HourlyRate.String(),
+			YearlyRate: lendingResp.PredictedRate.YearlyRate.String(),
+		}
+		if r.GetBorrowRates {
+			resp.PredictedRate.HourlyBorrowRate = lendingResp.PredictedRate.HourlyBorrowRate.String()
+			resp.PredictedRate.YearlyBorrowRate = lendingResp.PredictedRate.YearlyBorrowRate.String()
+		}
+	}
+	if r.IncludeAllRates {
+		for i := range lendingResp.Rates {
```
Comment (thrasher-) [inline engine/rpcserver.go:4901]: Can add a `resp.Rates = make([]*gctrpc.MarginRate, len(lendingResp.Rates))` here and then `resp.Rates[i] = rate` at the end of the for loop

File: engine/rpcserver.go [hunk 4]
```diff
@@ -4749,3 +4750,180 @@ func (s *RPCServer) GetTechnicalAnalysis(ctx context.Context, r *gctrpc.GetTechn
...
+			return nil, err
+		}
+
+		if request.TakeFeeRate.LessThanOrEqual(decimal.Zero) {
+			return nil, fmt.Errorf("%w for offline calculations", common.ErrCannotCalculateOffline)
+		}
+		if len(r.Rates) == 0 {
+			return nil, fmt.Errorf("%w for offline calculations", common.ErrCannotCalculateOffline)
+		}
+		for i := range r.Rates {
```
Comment (thrasher-) [inline engine/rpcserver.go:4829]: Can add `request.Rates = make([]margin.Rate, len(r.Rates))` and `request.Rates[i] = offlineRate` at the end of the for loop

**Review by thrasher- — CHANGES_REQUESTED** [review]
Very nice PR! Just a few minor nits but after that, it's a tACK from me:

image

**Review by thrasher- — APPROVED** [review]
tACK, thanks for making those changes!

**Pull request merged**